### PR TITLE
Adding Bokinfo coverimages support. (Reworked)

### PR DESCRIPTION
**Pull request by JacobSandin**
Support for using https://www.bokinfo.se as a source for coverimages.

### CODE REVIEW

File: module/VuFind/src/VuFind/Content/Covers/Bokinfo.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: This comment isn't accurate.

File: module/VuFind/src/VuFind/Content/Covers/Bokinfo.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: These copyright notices probably aren't right either.

File: module/VuFind/src/VuFind/Content/Covers/Bokinfo.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: This comment should also be adjusted to match whatever text you choose above.

File: module/VuFind/src/VuFind/Content/Covers/Bokinfo.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: Feel free to give yourself author credit.

File: module/VuFind/src/VuFind/Content/Covers/Bokinfo.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: You could reformat this to:

`// TODO: confirm whether Bokinfo allows caching before uncommenting`

(Which makes it a little more clear what the next step is here).

File: module/VuFind/src/VuFind/Content/Covers/Bokinfo.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: Were you going to set the method to HEAD here, or did you find that it is not necessary/doesn't work?

File: module/VuFind/src/VuFind/Content/Covers/PluginManager.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Content/Covers/PluginManager.php]: I try to keep the configuration lists in alphabetical order; do you mind moving the factory and alias lines up?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for all the progress -- just a few more questions and comments on the latest revision!

File: module/VuFind/src/VuFind/Content/Covers/Bokinfo.php
Comment (JacobSandin) [reply module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: It actually worked fine without, so therefore I removed it.

File: module/VuFind/src/VuFind/Content/Covers/PluginManager.php
Comment (JacobSandin) [reply module/VuFind/src/VuFind/Content/Covers/PluginManager.php]: Do not mind at all.

File: module/VuFind/src/VuFind/Content/Covers/Bokinfo.php
Comment (JacobSandin) [reply module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: They confirmed that it was ok.

File: module/VuFind/src/VuFind/Content/Covers/Bokinfo.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: This mentions Summon, which is not correct; I think you should just copy the "Plugin for Bokinfo coverimages" line from above for consistency.

(I would have done this myself to save you the trouble, but it looks like I don't have write access to this fork).

File: module/VuFind/src/VuFind/Content/Covers/Bokinfo.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: Maybe this should be `if (empty($key))` rather than `if (!isset($key))`. By virtue of being a required function parameter, `$key` is always going to be set to something... but if it's an empty string, that's going to be a problem. :-)

File: module/VuFind/src/VuFind/Content/Covers/Bokinfo.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: You don't need the double quotes here if you're just returning the variable unmodified -- they serve no purpose in this context.

File: module/VuFind/src/VuFind/Content/Covers/Bokinfo.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: I'm curious, though, if this is causing VuFind to effectively download every image twice -- once to check if it's valid, and again to cache/display it. Is there a reason that you have to do this? Are you unable to simply trust the results of the API call?

**Review by demiankatz — COMMENTED** [review]
Thanks -- we're getting close to finishing this (though if you want to be really thorough, a unit test would be great... but that can wait if it's an unreasonable hurdle at this stage). Just a few more minor comments below...

File: module/VuFind/src/VuFind/Content/Covers/Bokinfo.php
Comment (JacobSandin) [reply module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: I do not mind learning to get things right. However, how would I give that access? I guess it gets removed on  a fork?

Comment (demiankatz) [reply module/VuFind/src/VuFind/Content/Covers/Bokinfo.php]: I received your invitation, so I see you have figured this out. :-) I believe there is also a checkbox when you create a pull request that lets you give permission to project owners to modify your PR branch without giving access to your entire repository, though there may be repository-level settings that impact whether or not the option appears. In any case, I think we're in good shape now; thanks!

**Comment by demiankatz** [comment]
Thanks, @JacobSandin! I've added a comment to the class indicating that we checked about cache permissions (just in case we forget in the future, I wanted to capture the fact that we had done this).

I think this can be merged, though if you'd like me to add a unit test to show you how that's done, I'd be happy to help if you can provide an example XML response from the service to use as a test fixture.

**Comment by demiankatz** [comment]
Thanks for sharing test data, @JacobSandin! I've added a test and made a few very small edits. Please review my work, and let me know if you have any questions. If you are happy with this, I believe it is ready to be merged!

**Comment by JacobSandin** [comment]
I is all good for me @demiankatz, will take me a bit longer to understand the testing and mocking. However I will be happy to get the code added.

**Review by demiankatz — APPROVED** [review]

**Comment by demiankatz** [comment]
Thanks, @JacobSandin -- merging now!

**Pull request merged**
